packages/prettier-plugin-java/src/printers/printer-utils.js

```diff
@@ -443,6 +443,67 @@ function isShiftOperator(tokens, index) {
   return "none";
 }
 
+function retrieveNodesToken(ctx) {
+  const tokens = retrieveNodesTokenRec(ctx);
+  tokens.sort((token1, token2) => {
+    return token1.startOffset - token2.startOffset;
+  });
+  return tokens;
+}
+
+function retrieveNodesTokenRec(ctx) {
+  const tokens = [];
+  if (ctx && ctx.hasOwnProperty("image") && ctx.tokenType) {
+    if (ctx.leadingComments) {
+      tokens.push(...ctx.leadingComments);
+    }
+    tokens.push(ctx);
+    if (ctx.trailingComments) {
+      tokens.push(...ctx.trailingComments);
+    }
+    return tokens;
+  }
+  Object.keys(ctx.children).forEach(child => {
+    ctx.children[child].forEach(subctx => {
+      tokens.push(...retrieveNodesTokenRec(subctx));
+    });
+  });
+  return tokens;
+}
+
+function buildOriginalText(tokens) {
```

Comment:
Building the original text may be error prone.
Could we get a substring of the original text given start/end offsets of the CSTNode
and slice it into the new "formatted" output we are creating instead?

Reply:
It could work but the start/end offsets on CSTNode only consider the non comment token so if we have 
```java
// prettier-ignore
int a = 1; // my int
```
it will slice and give
```java
int a = 1;
```
Thus removing the comments.

Reply:
A workaround would be to retrieve all the tokens on the nodes and check if there is a leading/trailing comment. WDYT ?

Reply:
> A workaround would be to retrieve all the tokens on the nodes and check if there is a leading/trailing comment. WDYT ?

In general this still seems less error prone.
- Why do we need to check all Tokens on Node though? won't the first and last tokens be enough?

Reply:
Yes it will, this is what I am doing right now https://github.com/jhipster/prettier-java/pull/221/files#diff-21eac804fc21eb2d2bb5849e0cf5e7eeR474. I check the first and last token.

Reply:
sounds good!